Name: Muley, Tushar

Assignment: Week 7 & 8 Exercises

Date: February 7, 2021

# Exercise 1 Chapter 7: 


In [1]:
#import libraries
import numpy as np
import pandas as pd

In [2]:
#Path to location and name of file.
loc = 'candyhierarchy2017.xlsx' 
#Read and load file into dataframe
df_candy=pd.read_excel(loc, index_col=None, header=0)

In [144]:
df_candy

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
0        90258773            NaN        NaN     NaN         NaN   
1        90272821             No       Male      44        USA    
2        90272829            NaN       Male      49         USA   
3        90272840             No       Male      40          us   
4        90272841             No       Male      23         usa   
...           ...            ...        ...     ...         ...   
2455     90314359             No       Male      24         USA   
2456     90314580             No     Female      33         USA   
2457     90314634             No     Female      26        USA    
2458     90314658             No       Male      58         Usa   
2459     90314802             No     Female      66         usa   

     Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
0                                 NaN                NaN   
1                                  NM                MEH   
2                            Virginia                NaN   
3                                  or                MEH   
4                            exton pa                JOY   
...                               ...                ...   
2455                               MD                JOY   
2456                         New York                MEH   
2457                        Tennessee                MEH   
2458                   North Carolina                NaN   
2459                     Pennsylvania            DESPAIR   

     Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                   NaN                                       
1                                               DESPAIR                                       
2                                                   NaN                                       
3                                               DESPAIR                                       
4                                               DESPAIR                                       
...                                                 ...                                       
2455                                            DESPAIR                                       
2456                                            DESPAIR                                       
2457                                            DESPAIR                                       
2458                                                NaN                                       
2459                                            DESPAIR                                       

     Q6 | Any full-sized candy bar Q6 | Black Jacks  ... Q8: DESPAIR OTHER  \
0                              NaN              NaN  ...               NaN   
1                              JOY              MEH  ...               NaN   
2                              NaN              NaN  ...               NaN   
3                              JOY              MEH  ...               NaN   
4                              JOY          DESPAIR  ...               NaN   
...                            ...              ...  ...               ...   
2455                           MEH          DESPAIR  ...  Fruit Stripe Gum   
2456                           JOY              NaN  ...            Capers   
2457                           JOY          DESPAIR  ...               NaN   
2458                           NaN              NaN  ...               NaN   
2459                           JOY          DESPAIR  ...               NaN   

                                     Q9: OTHER COMMENTS      Q10: DRESS  \
0                                                   NaN             NaN   
1     Bottom line is Twix is really the only candy w...  White and gold   
2                                                   NaN             NaN   
3                                Raisins can go to hell  White and gold   
4                          

Remove out duplicates

In [145]:
#Checking the size of the datafame
df_candy.shape


(2460, 120)

In [146]:
#Sorting the dataframe
df_candy_sorted = df_candy.sort_values(by=['Internal ID'])

In [147]:
#Checking for duplicates
df_candy_sorted.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
2455    False
2456    False
2457    False
2458    False
2459    False
Length: 2460, dtype: bool

In [149]:
#Try droping duplicate data
df_candy_removed_dups = df_candy_sorted.drop_duplicates()

In [150]:
df_candy_removed_dups.shape

(2460, 120)

In [151]:
#Removed duplicates but specific column name
df_candy_removed_dups=df_candy_sorted.drop_duplicates(['Internal ID'])

In [153]:
df_candy_removed_dups.shape

(2460, 120)

The data does not appear to have no duplicates.

Filter out missing data

In [155]:
#Check to see which columns are missing a lot of data.
for c in df_candy.columns:
    miss=df_candy[c].isnull().sum()
    if miss > 0:
        print('{} has {} missing values'.format(c,miss))
    else:
        print('{} has NO missing values'.format(c))

Internal ID has NO missing values
Q1: GOING OUT? has 110 missing values
Q2: GENDER has 41 missing values
Q3: AGE has 84 missing values
Q4: COUNTRY has 64 missing values
Q5: STATE, PROVINCE, COUNTY, ETC has 100 missing values
Q6 | 100 Grand Bar has 747 missing values
Q6 | Anonymous brown globs that come in black and orange wrappers	(a.k.a. Mary Janes) has 734 missing values
Q6 | Any full-sized candy bar has 672 missing values
Q6 | Black Jacks has 958 missing values
Q6 | Bonkers (the candy) has 992 missing values
Q6 | Bonkers (the board game) has 1006 missing values
Q6 | Bottle Caps has 765 missing values
Q6 | Box'o'Raisins has 688 missing values
Q6 | Broken glow stick has 706 missing values
Q6 | Butterfinger has 682 missing values
Q6 | Cadbury Creme Eggs has 683 missing values
Q6 | Candy Corn has 678 missing values
Q6 | Candy that is clearly just the stuff given out for free at restaurants has 691 missing values
Q6 | Caramellos has 752 missing values
Q6 | Cash, or other forms of legal t

It appears Q12 has a lot of missing data. With out total row count of 2,460 missing a vary large part of the answers.
Q12: MEDIA [Daily Dish] has 2375 missing values
Q12: MEDIA [Science] has 1098 missing values
Q12: MEDIA [ESPN] has 2361 missing values
Q12: MEDIA [Yahoo] has 2393 missing values

In [156]:
#Drop rows that are all null values
df_candy.dropna(how='all')

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
0        90258773            NaN        NaN     NaN         NaN   
1        90272821             No       Male      44        USA    
2        90272829            NaN       Male      49         USA   
3        90272840             No       Male      40          us   
4        90272841             No       Male      23         usa   
...           ...            ...        ...     ...         ...   
2455     90314359             No       Male      24         USA   
2456     90314580             No     Female      33         USA   
2457     90314634             No     Female      26        USA    
2458     90314658             No       Male      58         Usa   
2459     90314802             No     Female      66         usa   

     Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
0                                 NaN                NaN   
1                                  NM                MEH   
2                            Virginia                NaN   
3                                  or                MEH   
4                            exton pa                JOY   
...                               ...                ...   
2455                               MD                JOY   
2456                         New York                MEH   
2457                        Tennessee                MEH   
2458                   North Carolina                NaN   
2459                     Pennsylvania            DESPAIR   

     Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                   NaN                                       
1                                               DESPAIR                                       
2                                                   NaN                                       
3                                               DESPAIR                                       
4                                               DESPAIR                                       
...                                                 ...                                       
2455                                            DESPAIR                                       
2456                                            DESPAIR                                       
2457                                            DESPAIR                                       
2458                                                NaN                                       
2459                                            DESPAIR                                       

     Q6 | Any full-sized candy bar Q6 | Black Jacks  ... Q8: DESPAIR OTHER  \
0                              NaN              NaN  ...               NaN   
1                              JOY              MEH  ...               NaN   
2                              NaN              NaN  ...               NaN   
3                              JOY              MEH  ...               NaN   
4                              JOY          DESPAIR  ...               NaN   
...                            ...              ...  ...               ...   
2455                           MEH          DESPAIR  ...  Fruit Stripe Gum   
2456                           JOY              NaN  ...            Capers   
2457                           JOY          DESPAIR  ...               NaN   
2458                           NaN              NaN  ...               NaN   
2459                           JOY          DESPAIR  ...               NaN   

                                     Q9: OTHER COMMENTS      Q10: DRESS  \
0                                                   NaN             NaN   
1     Bottom line is Twix is really the only candy w...  White and gold   
2                                                   NaN             NaN   
3                                Raisins can go to hell  White and gold   
4                          

In [157]:
#Only keep data with ages. Reverse of dropping.
df_candy=df_candy[df_candy['Q3: AGE'].notna()]


In [158]:
df_candy.shape

(2376, 120)

In [159]:
df_candy

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
1        90272821             No       Male      44        USA    
2        90272829            NaN       Male      49         USA   
3        90272840             No       Male      40          us   
4        90272841             No       Male      23         usa   
6        90272853             No       Male      53         usa   
...           ...            ...        ...     ...         ...   
2455     90314359             No       Male      24         USA   
2456     90314580             No     Female      33         USA   
2457     90314634             No     Female      26        USA    
2458     90314658             No       Male      58         Usa   
2459     90314802             No     Female      66         usa   

     Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
1                                  NM                MEH   
2                            Virginia                NaN   
3                                  or                MEH   
4                            exton pa                JOY   
6                           Colorado                 NaN   
...                               ...                ...   
2455                               MD                JOY   
2456                         New York                MEH   
2457                        Tennessee                MEH   
2458                   North Carolina                NaN   
2459                     Pennsylvania            DESPAIR   

     Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
1                                               DESPAIR                                       
2                                                   NaN                                       
3                                               DESPAIR                                       
4                                               DESPAIR                                       
6                                                   NaN                                       
...                                                 ...                                       
2455                                            DESPAIR                                       
2456                                            DESPAIR                                       
2457                                            DESPAIR                                       
2458                                                NaN                                       
2459                                            DESPAIR                                       

     Q6 | Any full-sized candy bar Q6 | Black Jacks  ... Q8: DESPAIR OTHER  \
1                              JOY              MEH  ...               NaN   
2                              NaN              NaN  ...               NaN   
3                              JOY              MEH  ...               NaN   
4                              JOY          DESPAIR  ...               NaN   
6                              NaN              NaN  ...               NaN   
...                            ...              ...  ...               ...   
2455                           MEH          DESPAIR  ...  Fruit Stripe Gum   
2456                           JOY              NaN  ...            Capers   
2457                           JOY          DESPAIR  ...               NaN   
2458                           NaN              NaN  ...               NaN   
2459                           JOY          DESPAIR  ...               NaN   

                                     Q9: OTHER COMMENTS      Q10: DRESS  \
1     Bottom line is Twix is really the only candy w...  White and gold   
2                                                   NaN             NaN   
3                                Raisins can go to hell  White and gold   
4                                                   NaN  White and gold   
6                          

In [160]:
#Drop columns that are all null
df_candy.dropna(axis=1, how='all')

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
1        90272821             No       Male      44        USA    
2        90272829            NaN       Male      49         USA   
3        90272840             No       Male      40          us   
4        90272841             No       Male      23         usa   
6        90272853             No       Male      53         usa   
...           ...            ...        ...     ...         ...   
2455     90314359             No       Male      24         USA   
2456     90314580             No     Female      33         USA   
2457     90314634             No     Female      26        USA    
2458     90314658             No       Male      58         Usa   
2459     90314802             No     Female      66         usa   

     Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
1                                  NM                MEH   
2                            Virginia                NaN   
3                                  or                MEH   
4                            exton pa                JOY   
6                           Colorado                 NaN   
...                               ...                ...   
2455                               MD                JOY   
2456                         New York                MEH   
2457                        Tennessee                MEH   
2458                   North Carolina                NaN   
2459                     Pennsylvania            DESPAIR   

     Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
1                                               DESPAIR                                       
2                                                   NaN                                       
3                                               DESPAIR                                       
4                                               DESPAIR                                       
6                                                   NaN                                       
...                                                 ...                                       
2455                                            DESPAIR                                       
2456                                            DESPAIR                                       
2457                                            DESPAIR                                       
2458                                                NaN                                       
2459                                            DESPAIR                                       

     Q6 | Any full-sized candy bar Q6 | Black Jacks  ... Q8: DESPAIR OTHER  \
1                              JOY              MEH  ...               NaN   
2                              NaN              NaN  ...               NaN   
3                              JOY              MEH  ...               NaN   
4                              JOY          DESPAIR  ...               NaN   
6                              NaN              NaN  ...               NaN   
...                            ...              ...  ...               ...   
2455                           MEH          DESPAIR  ...  Fruit Stripe Gum   
2456                           JOY              NaN  ...            Capers   
2457                           JOY          DESPAIR  ...               NaN   
2458                           NaN              NaN  ...               NaN   
2459                           JOY          DESPAIR  ...               NaN   

                                     Q9: OTHER COMMENTS      Q10: DRESS  \
1     Bottom line is Twix is really the only candy w...  White and gold   
2                                                   NaN             NaN   
3                                Raisins can go to hell  White and gold   
4                                                   NaN  White and gold   
6                          

In [161]:
df_candy.columns

Index(['Internal ID', 'Q1: GOING OUT?', 'Q2: GENDER', 'Q3: AGE', 'Q4: COUNTRY',
       'Q5: STATE, PROVINCE, COUNTY, ETC', 'Q6 | 100 Grand Bar',
       'Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)',
       'Q6 | Any full-sized candy bar', 'Q6 | Black Jacks',
       ...
       'Q8: DESPAIR OTHER', 'Q9: OTHER COMMENTS', 'Q10: DRESS', 'Unnamed: 113',
       'Q11: DAY', 'Q12: MEDIA [Daily Dish]', 'Q12: MEDIA [Science]',
       'Q12: MEDIA [ESPN]', 'Q12: MEDIA [Yahoo]', 'Click Coordinates (x, y)'],
      dtype='object', length=120)

In [162]:
df_candy.dtypes

Internal ID                   int64
Q1: GOING OUT?               object
Q2: GENDER                   object
Q3: AGE                      object
Q4: COUNTRY                  object
                             ...   
Q12: MEDIA [Daily Dish]     float64
Q12: MEDIA [Science]        float64
Q12: MEDIA [ESPN]           float64
Q12: MEDIA [Yahoo]          float64
Click Coordinates (x, y)     object
Length: 120, dtype: object

In [3]:
#Replacing Values that are not numbers with NaN
df_candy=df_candy.replace(['Old enough','Many','?','no','45-55','hahahahaha','older than dirt','5u','Enough',
                          'See question 2','24-50','Over 50','46 Halloweens.','ancient','OLD','old','MY NAME JEFF',
                           'sixty-nine','59 on the day after Halloween','old enough','your mom',
                          'I can remember when Java was a cool new language','60+'], np.nan)

In [4]:
#Keep only none NA values
df_candy=df_candy[df_candy['Q3: AGE'].notna()]


In [5]:
#Change all string to ints in AGE column
df_candy['Q3: AGE'] = df_candy['Q3: AGE'].astype(int)


In [195]:
#Bing for the age
bins = [18, 25, 35, 45, 55, 65, 100]

In [196]:
#Create the bin for age in the candy dataframe
df_candy_bin = pd.cut(df_candy['Q3: AGE'], bins=bins)

In [197]:
#View the df_candy_bin dataframe
df_candy_bin

1        (35, 45]
2        (45, 55]
3        (35, 45]
4        (18, 25]
6        (45, 55]
          ...    
2455     (18, 25]
2456     (25, 35]
2457     (25, 35]
2458     (55, 65]
2459    (65, 100]
Name: Q3: AGE, Length: 2352, dtype: category
Categories (6, interval[int64]): [(18, 25] < (25, 35] < (35, 45] < (45, 55] < (55, 65] < (65, 100]]

In [198]:
#Value count the bins
pd.value_counts(df_candy_bin)

(35, 45]     796
(25, 35]     549
(45, 55]     538
(55, 65]     250
(18, 25]      84
(65, 100]     66
Name: Q3: AGE, dtype: int64

# Exercise 1 Chapter 8: 

In [6]:
#Copy the original df_candy dataframe to perform a combine merge
df_candy_fhalf = df_candy.copy()
df_candy_shalf = df_candy.copy()

In [4]:
df_candy_fhalf.columns

Index(['Internal ID', 'Q1: GOING OUT?', 'Q2: GENDER', 'Q3: AGE', 'Q4: COUNTRY',
       'Q5: STATE, PROVINCE, COUNTY, ETC', 'Q6 | 100 Grand Bar',
       'Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)',
       'Q6 | Any full-sized candy bar', 'Q6 | Black Jacks',
       ...
       'Q8: DESPAIR OTHER', 'Q9: OTHER COMMENTS', 'Q10: DRESS', 'Unnamed: 113',
       'Q11: DAY', 'Q12: MEDIA [Daily Dish]', 'Q12: MEDIA [Science]',
       'Q12: MEDIA [ESPN]', 'Q12: MEDIA [Yahoo]', 'Click Coordinates (x, y)'],
      dtype='object', length=120)

In [5]:
#Drop the everything Q6 back to create a dataset for merging
df_candy_fhalf.drop(df_candy_fhalf.iloc[:, 6:120], inplace = True, axis = 1) 

In [6]:
#Check the columns have been dropped
df_candy_fhalf

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
0        90258773            NaN        NaN     NaN         NaN   
1        90272821             No       Male      44        USA    
2        90272829            NaN       Male      49         USA   
3        90272840             No       Male      40          us   
4        90272841             No       Male      23         usa   
...           ...            ...        ...     ...         ...   
2455     90314359             No       Male      24         USA   
2456     90314580             No     Female      33         USA   
2457     90314634             No     Female      26        USA    
2458     90314658             No       Male      58         Usa   
2459     90314802             No     Female      66         usa   

     Q5: STATE, PROVINCE, COUNTY, ETC  
0                                 NaN  
1                                  NM  
2                            Virginia  
3                                  or  
4                            exton pa  
...                               ...  
2455                               MD  
2456                         New York  
2457                        Tennessee  
2458                   North Carolina  
2459                     Pennsylvania  

[2460 rows x 6 columns]

In [217]:
#Double check original DF was not changed.
df_candy.columns

Index(['Internal ID', 'Q1: GOING OUT?', 'Q2: GENDER', 'Q3: AGE', 'Q4: COUNTRY',
       'Q5: STATE, PROVINCE, COUNTY, ETC', 'Q6 | 100 Grand Bar',
       'Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)',
       'Q6 | Any full-sized candy bar', 'Q6 | Black Jacks',
       ...
       'Q8: DESPAIR OTHER', 'Q9: OTHER COMMENTS', 'Q10: DRESS', 'Unnamed: 113',
       'Q11: DAY', 'Q12: MEDIA [Daily Dish]', 'Q12: MEDIA [Science]',
       'Q12: MEDIA [ESPN]', 'Q12: MEDIA [Yahoo]', 'Click Coordinates (x, y)'],
      dtype='object', length=120)

In [225]:
#Drop first half of columns in the data set
df_candy_shalf.drop(df_candy_shalf.iloc[:, 1:6], inplace = True, axis = 1) 

In [226]:
df_candy_shalf

Internal ID Q6 | 100 Grand Bar  \
1        90272821                MEH   
2        90272829                NaN   
3        90272840                MEH   
4        90272841                JOY   
6        90272853                NaN   
...           ...                ...   
2455     90314359                JOY   
2456     90314580                MEH   
2457     90314634                MEH   
2458     90314658                NaN   
2459     90314802            DESPAIR   

     Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
1                                               DESPAIR                                       
2                                                   NaN                                       
3                                               DESPAIR                                       
4                                               DESPAIR                                       
6                                                   NaN                                       
...                                                 ...                                       
2455                                            DESPAIR                                       
2456                                            DESPAIR                                       
2457                                            DESPAIR                                       
2458                                                NaN                                       
2459                                            DESPAIR                                       

     Q6 | Any full-sized candy bar Q6 | Black Jacks Q6 | Bonkers (the candy)  \
1                              JOY              MEH                  DESPAIR   
2                              NaN              NaN                      NaN   
3                              JOY              MEH                      MEH   
4                              JOY          DESPAIR                      MEH   
6                              NaN              NaN                      NaN   
...                            ...              ...                      ...   
2455                           MEH          DESPAIR                  DESPAIR   
2456                           JOY              NaN                      NaN   
2457                           JOY          DESPAIR                      MEH   
2458                           NaN              NaN                      NaN   
2459                           JOY          DESPAIR                  DESPAIR   

     Q6 | Bonkers (the board game) Q6 | Bottle Caps Q6 | Box'o'Raisins  \
1                          DESPAIR          DESPAIR            DESPAIR   
2                              NaN              NaN                NaN   
3                          DESPAIR              MEH            DESPAIR   
4                          DESPAIR              MEH            DESPAIR   
6                              NaN              NaN                NaN   
...                            ...              ...                ...   
2455                           MEH              MEH            DESPAIR   
2456                           NaN              NaN            DESPAIR   
2457                           JOY          DESPAIR                MEH   
2458                           NaN              NaN                NaN   
2459                       DESPAIR          DESPAIR                MEH   

     Q6 | Broken glow stick  ... Q8: DESPAIR OTHER  \
1                   DESPAIR  ...               NaN   
2                       NaN  ...               NaN   
3                   DESPAIR  ...               NaN   
4                   DESPAIR  ...               NaN   
6                       NaN  ...               NaN   
...                     ...  ...               ...   
2455                DESPAIR  ...  Fruit Stripe Gum   
2456                DESPAIR  ...            Capers   
2457                    MEH  ...               N

In [230]:
#Merge the split data frame back together but with a twist the question Q6 are first and Q1-Q5 are last 
#using Internal ID'

df_candy_megered = pd.merge(df_candy_shalf, df_candy_fhalf, on='Internal ID')

In [231]:
#Review the columns to make sure the twist worked
df_candy_megered.columns

Index(['Internal ID', 'Q6 | 100 Grand Bar',
       'Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)',
       'Q6 | Any full-sized candy bar', 'Q6 | Black Jacks',
       'Q6 | Bonkers (the candy)', 'Q6 | Bonkers (the board game)',
       'Q6 | Bottle Caps', 'Q6 | Box'o'Raisins', 'Q6 | Broken glow stick',
       ...
       'Q12: MEDIA [Daily Dish]', 'Q12: MEDIA [Science]', 'Q12: MEDIA [ESPN]',
       'Q12: MEDIA [Yahoo]', 'Click Coordinates (x, y)', 'Q1: GOING OUT?',
       'Q2: GENDER', 'Q3: AGE', 'Q4: COUNTRY',
       'Q5: STATE, PROVINCE, COUNTY, ETC'],
      dtype='object', length=120)

In [232]:
#Review the data.
df_candy_megered

Internal ID Q6 | 100 Grand Bar  \
0        90272821                MEH   
1        90272829                NaN   
2        90272840                MEH   
3        90272841                JOY   
4        90272853                NaN   
...           ...                ...   
2347     90314359                JOY   
2348     90314580                MEH   
2349     90314634                MEH   
2350     90314658                NaN   
2351     90314802            DESPAIR   

     Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                               DESPAIR                                       
1                                                   NaN                                       
2                                               DESPAIR                                       
3                                               DESPAIR                                       
4                                                   NaN                                       
...                                                 ...                                       
2347                                            DESPAIR                                       
2348                                            DESPAIR                                       
2349                                            DESPAIR                                       
2350                                                NaN                                       
2351                                            DESPAIR                                       

     Q6 | Any full-sized candy bar Q6 | Black Jacks Q6 | Bonkers (the candy)  \
0                              JOY              MEH                  DESPAIR   
1                              NaN              NaN                      NaN   
2                              JOY              MEH                      MEH   
3                              JOY          DESPAIR                      MEH   
4                              NaN              NaN                      NaN   
...                            ...              ...                      ...   
2347                           MEH          DESPAIR                  DESPAIR   
2348                           JOY              NaN                      NaN   
2349                           JOY          DESPAIR                      MEH   
2350                           NaN              NaN                      NaN   
2351                           JOY          DESPAIR                  DESPAIR   

     Q6 | Bonkers (the board game) Q6 | Bottle Caps Q6 | Box'o'Raisins  \
0                          DESPAIR          DESPAIR            DESPAIR   
1                              NaN              NaN                NaN   
2                          DESPAIR              MEH            DESPAIR   
3                          DESPAIR              MEH            DESPAIR   
4                              NaN              NaN                NaN   
...                            ...              ...                ...   
2347                           MEH              MEH            DESPAIR   
2348                           NaN              NaN            DESPAIR   
2349                           JOY          DESPAIR                MEH   
2350                           NaN              NaN                NaN   
2351                       DESPAIR          DESPAIR                MEH   

     Q6 | Broken glow stick  ... Q12: MEDIA [Daily Dish] Q12: MEDIA [Science]  \
0                   DESPAIR  ...                     NaN                  1.0   
1                       NaN  ...                     NaN                  NaN   
2                   DESPAIR  ...                     NaN                  1.0   
3                   DESPAIR  ...                     NaN                  1.0   
4                       NaN  ...                     NaN                  NaN   
...                     ...  ...                

In [7]:
#Copy over Internal ID and Age into two new dataframes to perform a combine
df_candy_1 = df_candy[['Internal ID','Q3: AGE']].head(100)

df_candy_2 = df_candy[['Internal ID','Q3: AGE']].head(100)

In [8]:
#Check the shape of df_candy_1. df_candy_2 should be the same shape
df_candy_1.shape

(100, 2)

In [10]:
#Review the df_candy_1
df_candy_1

Internal ID Q3: AGE
0      90258773     NaN
1      90272821      44
2      90272829      49
3      90272840      40
4      90272841      23
..          ...     ...
95     90273049      50
96     90273050      64
97     90273051      42
98     90273053      35
99     90273055      44

[100 rows x 2 columns]

In [29]:
#Rename the columns
df_candy_1.columns = ['Internal_ID', 'age']
df_candy_2.columns = ['Internal_ID', 'age']


In [30]:
#check changes
df_candy_1.columns

Index(['Internal_ID', 'age'], dtype='object')

In [31]:
#Check changes
df_candy_2.columns

Index(['Internal_ID', 'age'], dtype='object')

In [37]:
df_candy_1['age'] = df_candy_1['age'].astype(float)

In [45]:
#Adding in nan for everything that is age 44. This allow me to do a combine with df_candy_2 to show combine with
#where and combine_first.
#df_candy_1 = df_candy_1.replace(44,np.nan)

df_candy_1[(df_candy_1>=44) & (df_candy_1<=60)] = np.nan

In [46]:
df_candy_1

Internal_ID   age
0      90258773   NaN
1      90272821   NaN
2      90272829   NaN
3      90272840  40.0
4      90272841  23.0
..          ...   ...
95     90273049   NaN
96     90273050  64.0
97     90273051  42.0
98     90273053  35.0
99     90273055   NaN

[100 rows x 2 columns]

In [15]:
#View data type of df_candy_2 before change of data type
df_candy_2.dtypes

Internal ID     int64
Q3: AGE        object
dtype: object

In [253]:
#View of df_candy_2 before change of data type
df_candy_2

Internal ID  Q3: AGE
1       90272821       44
2       90272829       49
3       90272840       40
4       90272841       23
6       90272853       53
..           ...      ...
98      90273053       35
99      90273055       44
101     90273057       44
102     90273058       42
103     90273059       43

[100 rows x 2 columns]

In [16]:
#Change the data type from object to float to a combine can be down with df_candy_1.
df_candy_2['Q3: AGE'] = df_candy_2['Q3: AGE'].astype(float)

In [17]:
#Check data type after change to confirm
df_candy_2.dtypes

Internal ID      int64
Q3: AGE        float64
dtype: object

In [18]:
#Review changes to data
df_candy_2

Internal ID  Q3: AGE
0      90258773      NaN
1      90272821     44.0
2      90272829     49.0
3      90272840     40.0
4      90272841     23.0
..          ...      ...
95     90273049     50.0
96     90273050     64.0
97     90273051     42.0
98     90273053     35.0
99     90273055     44.0

[100 rows x 2 columns]

In [48]:
#Using combine_first to patch 
df_candy_1.combine_first(df_candy_2)

Internal_ID   age
0      90258773   NaN
1      90272821  44.0
2      90272829  49.0
3      90272840  40.0
4      90272841  23.0
..          ...   ...
95     90273049  50.0
96     90273050  64.0
97     90273051  42.0
98     90273053  35.0
99     90273055  44.0

[100 rows x 2 columns]

Before combine 	
    Internal_ID 	age
0 	90258773 	NaN
1 	90272821 	NaN
2 	90272829 	NaN
3 	90272840 	40.0
4 	90272841 	23.0
... 	... 	...
95 	90273049 	NaN
96 	90273050 	64.0
97 	90273051 	42.0
98 	90273053 	35.0
99 	90273055 	NaN

After combine most matching NaN got replaced with number from df_candy_2 that were available like Internal id (90272821, 90272829, 90273049, 90273055)
 	Internal_ID 	age
0 	90258773 	NaN
1 	90272821 	44.0
2 	90272829 	49.0
3 	90272840 	40.0
4 	90272841 	23.0
... 	... 	...
95 	90273049 	50.0
96 	90273050 	64.0
97 	90273051 	42.0
98 	90273053 	35.0
99 	90273055 	44.0

In [40]:
#Load candy file with dates
loc2 = 'BOING-BOING-CANDY-HIERARCHY-2016-SURVEY-Responses.xlsx' 
#Read and load file into dataframe
df_boing=pd.read_excel(loc2, index_col=None, header=0)

In [41]:
df_boing

Timestamp  \
0    2016-10-24 05:09:23.033   
1    2016-10-24 05:09:54.798   
2    2016-10-24 05:13:06.734   
3    2016-10-24 05:14:17.192   
4    2016-10-24 05:14:24.625   
...                      ...   
1254 2016-10-29 16:53:52.516   
1255 2016-10-30 06:53:54.735   
1256 2016-10-30 11:06:10.827   
1257 2016-10-30 16:07:26.539   
1258 2016-10-30 17:06:45.660   

     Are you going actually going trick or treating yourself? Your gender:  \
0                                                    No               Male   
1                                                    No               Male   
2                                                    No             Female   
3                                                    No               Male   
4                                                   Yes               Male   
...                                                 ...                ...   
1254                                                 No             Female   
1255                                                 No               Male   
1256                                                 No               Male   
1257                                                 No               Male   
1258                                                Yes             Female   

     How old are you? Which country do you live in?  \
0                  22                        Canada   
1                  45                           usa   
2                  48                            US   
3                  57                           usa   
4                  42                           USA   
...               ...                           ...   
1254               52                           USA   
1255               33                 united states   
1256              NaN                           NaN   
1257               48                        canada   
1258               44                            Us   

     Which state, province, county do you live in?  [100 Grand Bar]  \
0                                          Ontario              JOY   
1                                               il              MEH   
2                                         Colorado              JOY   
3                                               il              JOY   
4                                     South Dakota              MEH   
...                                            ...              ...   
1254                                            TX              JOY   
1255                                     minnesota              JOY   
1256                                           NaN              JOY   
1257                                            BC              NaN   
1258                                            Nh              JOY   

      [Anonymous brown globs that come in black and orange wrappers]  \
0                                               DESPAIR                
1                                                   MEH                
2                                               DESPAIR                
3                                                   MEH                
4                                               DESPAIR                
...                                                 ...                
1254                                            DESPAIR                
1255                                            DESPAIR                
1256                                                MEH                
1257                                            DESPAIR                
1258                                                MEH                

      [Any full-sized candy bar]  [Black Jacks]  ...  \
0                            JOY            MEH  ...   
1                            JOY            JOY  ...   
2                            JOY            MEH  ...   
3                            JOY            MEH  ...   
4                            JOY      

In [42]:
#Reduce the size by removing some columns to perform Pivot
df_small_boing= df_boing.drop(df_boing.iloc[:, 5:123], axis = 1) 

In [43]:
#Drop one additional column
df_small_boing.drop(df_small_boing.iloc[:, 1:2], inplace=True, axis = 1) 

In [44]:
#View the data
df_small_boing

Timestamp Your gender: How old are you?  \
0    2016-10-24 05:09:23.033         Male               22   
1    2016-10-24 05:09:54.798         Male               45   
2    2016-10-24 05:13:06.734       Female               48   
3    2016-10-24 05:14:17.192         Male               57   
4    2016-10-24 05:14:24.625         Male               42   
...                      ...          ...              ...   
1254 2016-10-29 16:53:52.516       Female               52   
1255 2016-10-30 06:53:54.735         Male               33   
1256 2016-10-30 11:06:10.827         Male              NaN   
1257 2016-10-30 16:07:26.539         Male               48   
1258 2016-10-30 17:06:45.660       Female               44   

     Which country do you live in?  
0                           Canada  
1                              usa  
2                               US  
3                              usa  
4                              USA  
...                            ...  
1254                           USA  
1255                 united states  
1256                           NaN  
1257                        canada  
1258                            Us  

[1259 rows x 4 columns]

In [45]:
#Review the columns
df_small_boing.columns

Index(['Timestamp', 'Your gender:', 'How old are you?',
       'Which country do you live in?'],
      dtype='object')

In [46]:
#Rename the columns
df_small_boing.columns = ['timestamp', 'gender','age','country']

In [47]:
#Confirm the columns
df_small_boing.columns

Index(['timestamp', 'gender', 'age', 'country'], dtype='object')

In [120]:
#Pivot the data.
df_pivoted_boing = df_small_boing.pivot('timestamp','gender')

In [118]:
#View the pivoted data.
df_pivoted_boing

gender                   NaN Female I'd rather not say Male Other
timestamp                                                        
2016-10-24 05:09:23.033  NaN    NaN                NaN   22   NaN
2016-10-24 05:09:54.798  NaN    NaN                NaN   45   NaN
2016-10-24 05:13:06.734  NaN     48                NaN  NaN   NaN
2016-10-24 05:14:17.192  NaN    NaN                NaN   57   NaN
2016-10-24 05:14:24.625  NaN    NaN                NaN   42   NaN
...                      ...    ...                ...  ...   ...
2016-10-29 16:53:52.516  NaN     52                NaN  NaN   NaN
2016-10-30 06:53:54.735  NaN    NaN                NaN   33   NaN
2016-10-30 11:06:10.827  NaN    NaN                NaN  NaN   NaN
2016-10-30 16:07:26.539  NaN    NaN                NaN   48   NaN
2016-10-30 17:06:45.660  NaN     44                NaN  NaN   NaN

[1259 rows x 5 columns]

Note: Interesting Gender across the top and only filled in where an age was provided.

In [124]:
#Pivot with country
#Pivot the data.
df_pivoted_boing_ctry = df_small_boing.pivot('timestamp','country')

In [128]:
#Just wanted to see what happens when you pivot with country
df_pivoted_boing_ctry.dropna(axis=1,how='all')


gender                                               \
country                    NaN   30   44   45   47   51   54  United States   
timestamp                                                                     
2016-10-24 05:09:23.033    NaN  NaN  NaN  NaN  NaN  NaN  NaN            NaN   
2016-10-24 05:09:54.798    NaN  NaN  NaN  NaN  NaN  NaN  NaN            NaN   
2016-10-24 05:13:06.734    NaN  NaN  NaN  NaN  NaN  NaN  NaN            NaN   
2016-10-24 05:14:17.192    NaN  NaN  NaN  NaN  NaN  NaN  NaN            NaN   
2016-10-24 05:14:24.625    NaN  NaN  NaN  NaN  NaN  NaN  NaN            NaN   
...                        ...  ...  ...  ...  ...  ...  ...            ...   
2016-10-29 16:53:52.516    NaN  NaN  NaN  NaN  NaN  NaN  NaN            NaN   
2016-10-30 06:53:54.735    NaN  NaN  NaN  NaN  NaN  NaN  NaN            NaN   
2016-10-30 11:06:10.827   Male  NaN  NaN  NaN  NaN  NaN  NaN            NaN   
2016-10-30 16:07:26.539    NaN  NaN  NaN  NaN  NaN  NaN  NaN            NaN   
2016-10-30 17:06:45.660    NaN  NaN  NaN  NaN  NaN  NaN  NaN            NaN   

                                                                        ...  \
country                 A tropical island south of the equator America  ...   
timestamp                                                               ...   
2016-10-24 05:09:23.033                                    NaN     NaN  ...   
2016-10-24 05:09:54.798                                    NaN     NaN  ...   
2016-10-24 05:13:06.734                                    NaN     NaN  ...   
2016-10-24 05:14:17.192                                    NaN     NaN  ...   
2016-10-24 05:14:24.625                                    NaN     NaN  ...   
...                                                        ...     ...  ...   
2016-10-29 16:53:52.516                                    NaN     NaN  ...   
2016-10-30 06:53:54.735                                    NaN     NaN  ...   
2016-10-30 11:06:10.827                                    NaN     NaN  ...   
2016-10-30 16:07:26.539                                    NaN     NaN  ...   
2016-10-30 17:06:45.660                                    NaN     NaN  ...   

                                                age                          \
country                 there isn't one for old men this one u.s.  uSA   uk   
timestamp                                                                     
2016-10-24 05:09:23.033                         NaN      NaN  NaN  NaN  NaN   
2016-10-24 05:09:54.798                         NaN      NaN  NaN  NaN  NaN   
2016-10-24 05:13:06.734                         NaN      NaN  NaN  NaN  NaN   
2016-10-24 05:14:17.192                         NaN      NaN  NaN  NaN  NaN   
2016-10-24 05:14:24.625                         NaN      NaN  NaN  NaN  NaN   
...                                             ...      ...  ...  ...  ...   
2016-10-29 16:53:52.516                         NaN      NaN  NaN  NaN  NaN   
2016-10-30 06:53:54.735                         NaN      NaN  NaN  NaN  NaN   
2016-10-30 11:06:10.827                         NaN      NaN  NaN  NaN  NaN   
2016-10-30 16:07:26.539                         NaN      NaN  NaN  NaN  NaN   
2016-10-30 17:06:45.660                         NaN      NaN  NaN  NaN  NaN   

                                                                               \
country                 united states united states  united states of america   
timestamp                                                                       
2016-10-24 05:09:23.033           NaN            NaN                      NaN   
2016-10-24 05:09:54.798           NaN            NaN                      NaN   
2016-10-24 05:13:06.734           NaN            NaN                      NaN   
2016-10-24 05:14:17.192           NaN            NaN                      NaN   
2016-10-24 05:14:24.625           NaN            NaN                      NaN   
...                               ...            ...   

In [33]:
#Just want to see if the dataframe changed.
df_small_boing

timestamp  gender  age        country
0    2016-10-24 05:09:23.033    Male   22         Canada
1    2016-10-24 05:09:54.798    Male   45            usa
2    2016-10-24 05:13:06.734  Female   48             US
3    2016-10-24 05:14:17.192    Male   57            usa
4    2016-10-24 05:14:24.625    Male   42            USA
...                      ...     ...  ...            ...
1254 2016-10-29 16:53:52.516  Female   52            USA
1255 2016-10-30 06:53:54.735    Male   33  united states
1256 2016-10-30 11:06:10.827    Male  NaN            NaN
1257 2016-10-30 16:07:26.539    Male   48         canada
1258 2016-10-30 17:06:45.660  Female   44             Us

[1259 rows x 4 columns]

# Exercise 1 Chapter 10: 

In [34]:
#Copy dataframe to use in groupby series
df_new_candy = df_small_boing.copy() 


In [35]:
#Do a few data clean up items
df_new_candy['country'] = df_new_candy['country'].str.replace(" ","")
df_new_candy.dropna(inplace=True)
df_new_candy['country'] = df_new_candy['country'].str.lower()
df_new_candy['gender'] = df_new_candy['gender'].str.lower()
df_new_candy=df_new_candy.replace(['us','u','america','unitedstateofamerica','unitedstates','usa!','usa! usa!',
                                  'usa! usa! usa!','usa!!!!!!','ussa','usausausa','usausausausa','unitedstate',
                                   'u.s.','u.s.a.','merica','murica'], 'usa')

In [36]:
df_new_candy=df_new_candy.replace(['Old enough','Many','?','no','45-55','hahahahaha','older than dirt','5u','Enough',
                          'See question 2','24-50','Over 50','46 Halloweens.','ancient','OLD','old','MY NAME JEFF',
                           'sixty-nine','59 on the day after Halloween','old enough','your mom',
                          'I can remember when Java was a cool new language','60+','Old enough to know better ',
                                  'As old as my tongue a few years older than my teeth','50s','0x2A',
                                  'Fifty.  Nine.  Ish.','Ancient','I remember the Nixon administration',
                                  'over retirement age','50+','55+','over 40','Hahahahahaha', 
                                   'Old','Older than i act','really old','blah','older than I want to be',
                                  'Not as old as you...','Never ask a woman that question.','old ',
                                  'Same as yo mama','Too old to trick or treat without it being creepy',
                                  'Old enough to not Trick or Treat.','49 11/12ths'], np.nan)

df_new_candy.dropna(inplace=True)


In [231]:
df_new_candy['age'] = df_new_candy['age'].astype(float)


In [232]:
#View the data
df_new_candy

timestamp  gender   age country
0    2016-10-24 05:09:23.033    male  22.0  canada
1    2016-10-24 05:09:54.798    male  45.0     usa
2    2016-10-24 05:13:06.734  female  48.0     usa
3    2016-10-24 05:14:17.192    male  57.0     usa
4    2016-10-24 05:14:24.625    male  42.0     usa
...                      ...     ...   ...     ...
1253 2016-10-29 14:47:43.907    male  54.0     usa
1254 2016-10-29 16:53:52.516  female  52.0     usa
1255 2016-10-30 06:53:54.735    male  33.0     usa
1257 2016-10-30 16:07:26.539    male  48.0  canada
1258 2016-10-30 17:06:45.660  female  44.0     usa

[1183 rows x 4 columns]

In [241]:
#Group by country
grouped = df_new_candy['age'].groupby(df_new_candy['country'])

In [242]:
grouped

In [243]:
#Group by country for the mean age of respondents
grouped.mean()

country
australia                 5.100000e+01
austria                   3.300000e+01
belgium                   4.200000e+01
brasil                    3.400000e+01
canada                    3.857391e+01
china                     6.700000e+01
croatia                   1.600000e+01
denial                    1.420000e+02
england                   3.550000e+01
españa                    6.000000e+01
eua                       3.800000e+01
finland                   3.400000e+01
france                    3.566667e+01
germany                   4.280000e+01
god'scountry              2.320000e+01
hungary                   3.500000e+01
japan                     4.500000e+01
kenya                     3.300000e+01
korea                     4.400000e+01
mexico                    4.300000e+01
netherlands               2.600000e+01
neverland                 7.800000e+01
newzealand                4.500000e+01
panama                    4.700000e+01
philippines               4.400000e+01
portugal         

In [245]:
#Different way to perform a grouping.
df_new_candy.groupby(['country']).mean()

age
country                             
australia               5.100000e+01
austria                 3.300000e+01
belgium                 4.200000e+01
brasil                  3.400000e+01
canada                  3.857391e+01
china                   6.700000e+01
croatia                 1.600000e+01
denial                  1.420000e+02
england                 3.550000e+01
españa                  6.000000e+01
eua                     3.800000e+01
finland                 3.400000e+01
france                  3.566667e+01
germany                 4.280000e+01
god'scountry            2.320000e+01
hungary                 3.500000e+01
japan                   4.500000e+01
kenya                   3.300000e+01
korea                   4.400000e+01
mexico                  4.300000e+01
netherlands             2.600000e+01
neverland               7.800000e+01
newzealand              4.500000e+01
panama                  4.700000e+01
philippines             4.400000e+01
portugal                5.800000e+01
somewhere               1.000000e+18
southkorea              2.500000e+01
sweden                  4.200000e+01
switzerland             3.500000e+01
thebestone-usa          3.600000e+01
thenetherlands          2.500000e+01
thereisn'toneforoldmen  5.700000e+01
therepublicofcascadia   4.500000e+01
theyooessofaaayyyyyy    4.600000e+01
uk                      4.175000e+01
unitedkindom            5.100000e+01
unitedkingdom           3.425000e+01
unitedsates             4.500000e+01
unitedstatesofamerica   3.384211e+01
unitedstetes            5.300000e+01
unitsstates             4.300000e+01
usa                     4.106165e+01

In [4]:
#Path to location and name of file. Load some term project data to perform grouping
loc = 'race_results_contractor.xlsx' 
#Read and load file into dataframe
df_contructor_results=pd.read_excel(loc, index_col=None, header=0)

In [5]:
#View the data
df_contructor_results

constructorResultsId                   race    contructor  points
0                  15641  Australian Grand Prix      Mercedes      22
1                  15642  Australian Grand Prix      Red Bull      20
2                  15643  Australian Grand Prix       McLaren      12
3                  15644  Australian Grand Prix       Renault       7
4                  15645  Australian Grand Prix   Force India       0
..                   ...                    ...           ...     ...
94                 15735     British Grand Prix       McLaren       4
95                 15736     British Grand Prix  Haas F1 Team       2
96                 15737     British Grand Prix      Williams       0
97                 15738     British Grand Prix    Toro Rosso       0
98                 15739     British Grand Prix        Sauber       0

[99 rows x 4 columns]

In [6]:
#Pivot the table on race and Contructor to view points.
df_contructor_results.pivot_table(index=['race', 'contructor'])

constructorResultsId  points
race                  contructor                                
Australian Grand Prix Force India                  15645       0
                      Haas F1 Team                 15649       0
                      McLaren                      15643      12
                      Mercedes                     15641      22
                      Red Bull                     15642      20
...                                                  ...     ...
Spanish Grand Prix    Red Bull                     15681      25
                      Renault                      15684       6
                      Sauber                       15687       1
                      Toro Rosso                   15689       0
                      Williams                     15688       0

[99 rows x 2 columns]

In [7]:
#Corsstab the data
pd.crosstab(df_contructor_results.race, df_contructor_results.contructor, margins=True)

contructor             Ferrari  Force India  Haas F1 Team  McLaren  Mercedes  \
race                                                                           
Australian Grand Prix        0            1             1        1         1   
Austrian Grand Prix          1            1             1        1         1   
Azerbaijan Grand Prix        1            1             1        1         1   
Bahrain Grand Prix           1            1             1        1         1   
British Grand Prix           1            1             1        1         1   
Canadian Grand Prix          1            1             1        1         1   
Chinese Grand Prix           1            1             1        1         1   
French Grand Prix            1            1             1        1         1   
Monaco Grand Prix            1            1             1        1         1   
Spanish Grand Prix           1            1             1        1         1   
All                          9           10            10       10        10   

contructor             Red Bull  Renault  Sauber  Toro Rosso  Williams  All  
race                                                                         
Australian Grand Prix         1        1       1           1         1    9  
Austrian Grand Prix           1        1       1           1         1   10  
Azerbaijan Grand Prix         1        1       1           1         1   10  
Bahrain Grand Prix            1        1       1           1         1   10  
British Grand Prix            1        1       1           1         1   10  
Canadian Grand Prix           1        1       1           1         1   10  
Chinese Grand Prix            1        1       1           1         1   10  
French Grand Prix             1        1       1           1         1   10  
Monaco Grand Prix             1        1       1           1         1   10  
Spanish Grand Prix            1        1       1           1         1   10  
All                          10       10      10          10        10   99

# Exercise 1 Chapter 11: 

In [48]:
#Copy datafram df_new_candy to another df so no further confusion happens
df_candy_chp11 = df_new_candy.copy()

In [49]:
#Review df_candy_chp11
df_candy_chp11.dtypes

timestamp    datetime64[ns]
gender               object
age                 float64
country              object
dtype: object

In [18]:
#import needed library
from datetime import timedelta

In [21]:
#Convert datetime to year only
df_year = df_candy_chp11['Timestamp'].dt.strftime('%Y')


In [261]:
#Print DF
df_year

0       2016
1       2016
2       2016
3       2016
4       2016
        ... 
1253    2016
1254    2016
1255    2016
1257    2016
1258    2016
Name: timestamp, Length: 1183, dtype: object

In [264]:
#Convert datetime to month only
df_month = df_candy_chp11['timestamp'].dt.strftime('%m')
df_month

0       10
1       10
2       10
3       10
4       10
        ..
1253    10
1254    10
1255    10
1257    10
1258    10
Name: timestamp, Length: 1183, dtype: object

In [265]:
#Convert datetime to day only
df_day = df_candy_chp11['timestamp'].dt.strftime('%d')
df_day

0       24
1       24
2       24
3       24
4       24
        ..
1253    29
1254    29
1255    30
1257    30
1258    30
Name: timestamp, Length: 1183, dtype: object

In [266]:
df_time = df_candy_chp11['timestamp'].dt.strftime('%H:%M:%S')
df_time

0       05:09:23
1       05:09:54
2       05:13:06
3       05:14:17
4       05:14:24
          ...   
1253    14:47:43
1254    16:53:52
1255    06:53:54
1257    16:07:26
1258    17:06:45
Name: timestamp, Length: 1183, dtype: object

In [50]:
#Convert datetime to string for everything
df_candy_chp11['strdatetime']=df_candy_chp11['timestamp'].dt.strftime('%m/%d/%Y, %H:%M:%S')
df_candy_chp11

timestamp  gender   age country           strdatetime
0    2016-10-24 05:09:23.033    male  22.0  canada  10/24/2016, 05:09:23
1    2016-10-24 05:09:54.798    male  45.0     usa  10/24/2016, 05:09:54
2    2016-10-24 05:13:06.734  female  48.0     usa  10/24/2016, 05:13:06
3    2016-10-24 05:14:17.192    male  57.0     usa  10/24/2016, 05:14:17
4    2016-10-24 05:14:24.625    male  42.0     usa  10/24/2016, 05:14:24
...                      ...     ...   ...     ...                   ...
1253 2016-10-29 14:47:43.907    male  54.0     usa  10/29/2016, 14:47:43
1254 2016-10-29 16:53:52.516  female  52.0     usa  10/29/2016, 16:53:52
1255 2016-10-30 06:53:54.735    male  33.0     usa  10/30/2016, 06:53:54
1257 2016-10-30 16:07:26.539    male  48.0  canada  10/30/2016, 16:07:26
1258 2016-10-30 17:06:45.660  female  44.0     usa  10/30/2016, 17:06:45

[1183 rows x 5 columns]

In [51]:
df_candy_chp11.dtypes

timestamp      datetime64[ns]
gender                 object
age                   float64
country                object
strdatetime            object
dtype: object

In [52]:
#Import library to convert string date time to datatime
from datetime import datetime

In [57]:
df_candy_chp11['datetime']=pd.to_datetime(df_candy_chp11['strdatetime'], format='%m/%d/%Y, %H:%M:%S')

In [58]:
df_candy_chp11

timestamp  gender   age country           strdatetime  \
0    2016-10-24 05:09:23.033    male  22.0  canada  10/24/2016, 05:09:23   
1    2016-10-24 05:09:54.798    male  45.0     usa  10/24/2016, 05:09:54   
2    2016-10-24 05:13:06.734  female  48.0     usa  10/24/2016, 05:13:06   
3    2016-10-24 05:14:17.192    male  57.0     usa  10/24/2016, 05:14:17   
4    2016-10-24 05:14:24.625    male  42.0     usa  10/24/2016, 05:14:24   
...                      ...     ...   ...     ...                   ...   
1253 2016-10-29 14:47:43.907    male  54.0     usa  10/29/2016, 14:47:43   
1254 2016-10-29 16:53:52.516  female  52.0     usa  10/29/2016, 16:53:52   
1255 2016-10-30 06:53:54.735    male  33.0     usa  10/30/2016, 06:53:54   
1257 2016-10-30 16:07:26.539    male  48.0  canada  10/30/2016, 16:07:26   
1258 2016-10-30 17:06:45.660  female  44.0     usa  10/30/2016, 17:06:45   

                datetime  
0    2016-10-24 05:09:23  
1    2016-10-24 05:09:54  
2    2016-10-24 05:13:06  
3    2016-10-24 05:14:17  
4    2016-10-24 05:14:24  
...                  ...  
1253 2016-10-29 14:47:43  
1254 2016-10-29 16:53:52  
1255 2016-10-30 06:53:54  
1257 2016-10-30 16:07:26  
1258 2016-10-30 17:06:45  

[1183 rows x 6 columns]

In [59]:
df_candy_chp11.dtypes

timestamp      datetime64[ns]
gender                 object
age                   float64
country                object
strdatetime            object
datetime       datetime64[ns]
dtype: object

In [95]:
#Perform date generations - Load csv file without dates
loc = 'candyhierarchy2017.xlsx' 
#Read and load file into dataframe
df_candy_2=pd.read_excel(loc, index_col=None, header=0)

In [96]:
df_candy_2.columns

Index(['Internal ID', 'Q1: GOING OUT?', 'Q2: GENDER', 'Q3: AGE', 'Q4: COUNTRY',
       'Q5: STATE, PROVINCE, COUNTY, ETC', 'Q6 | 100 Grand Bar',
       'Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)',
       'Q6 | Any full-sized candy bar', 'Q6 | Black Jacks',
       ...
       'Q8: DESPAIR OTHER', 'Q9: OTHER COMMENTS', 'Q10: DRESS', 'Unnamed: 113',
       'Q11: DAY', 'Q12: MEDIA [Daily Dish]', 'Q12: MEDIA [Science]',
       'Q12: MEDIA [ESPN]', 'Q12: MEDIA [Yahoo]', 'Click Coordinates (x, y)'],
      dtype='object', length=120)

In [98]:
#Check the shape of df
df_candy_2.shape

(2460, 120)

In [100]:
#Reduce the size of the df
df_candy_2.drop(df_candy_2.iloc[:, 5:120], axis = 1, inplace=True) 

In [101]:
#Check the shape of df
df_candy_2.shape

(2460, 5)

In [102]:
#Check the oolumns
df_candy_2.columns

Index(['Internal ID', 'Q1: GOING OUT?', 'Q2: GENDER', 'Q3: AGE',
       'Q4: COUNTRY'],
      dtype='object')

In [103]:
#Take a quick look
df_candy_2

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY
0        90258773            NaN        NaN     NaN         NaN
1        90272821             No       Male      44        USA 
2        90272829            NaN       Male      49         USA
3        90272840             No       Male      40          us
4        90272841             No       Male      23         usa
...           ...            ...        ...     ...         ...
2455     90314359             No       Male      24         USA
2456     90314580             No     Female      33         USA
2457     90314634             No     Female      26        USA 
2458     90314658             No       Male      58         Usa
2459     90314802             No     Female      66         usa

[2460 rows x 5 columns]

In [104]:
#Generate date range for the 2017 data that did not have a datetime.
df_candy_2['datetime'] = pd.date_range(start='10/24/2017', periods=len(df_candy_2), freq='12min')

In [105]:
df_candy_2

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
0        90258773            NaN        NaN     NaN         NaN   
1        90272821             No       Male      44        USA    
2        90272829            NaN       Male      49         USA   
3        90272840             No       Male      40          us   
4        90272841             No       Male      23         usa   
...           ...            ...        ...     ...         ...   
2455     90314359             No       Male      24         USA   
2456     90314580             No     Female      33         USA   
2457     90314634             No     Female      26        USA    
2458     90314658             No       Male      58         Usa   
2459     90314802             No     Female      66         usa   

                datetime  
0    2017-10-24 00:00:00  
1    2017-10-24 00:12:00  
2    2017-10-24 00:24:00  
3    2017-10-24 00:36:00  
4    2017-10-24 00:48:00  
...                  ...  
2455 2017-11-13 11:00:00  
2456 2017-11-13 11:12:00  
2457 2017-11-13 11:24:00  
2458 2017-11-13 11:36:00  
2459 2017-11-13 11:48:00  

[2460 rows x 6 columns]